vllm/v1/core/kv_cache_utils.py

```diff
@@ -173,37 +243,50 @@ def hash_block_tokens(parent_block_hash: Optional[int],
             if this is the first block.
         curr_block_token_ids: A tuple of token ids in the current
             block. The current block is assumed to be full.
+        extra_keys: Extra keys for the block.
 
     Returns:
         The hash value of the block and the token ids in the block.
         The entire tuple is used as the hash key of the block.
     """
     return BlockHashType(hash((parent_block_hash, *curr_block_token_ids)),
-                         curr_block_token_ids)
+                         curr_block_token_ids, extra_keys)
 
 
 def hash_request_tokens(block_size: int,
-                        token_ids: List[int]) -> List[BlockHashType]:
+                        request: Request) -> List[BlockHashType]:
     """Computes hash values of a chain of blocks given a sequence of
     token IDs. The hash value is used for prefix caching.
 
     Args:
         block_size: The size of each block.
-        token_ids: A sequence of token ids in the request.
+        request: The request object.
 
     Returns:
         The list of computed hash values.
     """
+    token_ids = request.all_token_ids
+    mm_positions, mm_hashes = request.mm_positions, request.mm_hashes
+    if mm_positions and len(mm_positions) != len(mm_hashes):
+        raise ValueError(
+            "The number of multi-modal positions and hashes must match.")
+
     ret = []
     parent_block_hash_value = None
+    curr_mm_idx = 0
     for start in range(0, len(token_ids), block_size):
         end = start + block_size
         block_token_ids = tuple(token_ids[start:end])
         # Do not hash the block if it is not full.
         if len(block_token_ids) < block_size:
             break
+
+        # Add extra keys if the block is a multi-modal block.
+        extra_keys, curr_mm_idx = generate_block_hash_extra_keys(
+            request, start, end, curr_mm_idx)
+
         block_hash = hash_block_tokens(parent_block_hash_value,
-                                       block_token_ids)
+                                       block_token_ids, extra_keys)
```

Comment:
This is mostly just a code style nit: What's your opinion on making `generate_block_hash_extra_keys` conditional in the loop?

If there is no `mm_positions` in the request (i.e, text only), we're still calling this function just to return `None` when we traverse the token ids. Since we already access `request.mm_positions` and `request_mm_hashes`, we can do something like
```python
    token_ids = request.all_token_ids
    mm_positions, mm_hashes = request.mm_positions, request.mm_hashes
    
    # This check is already in generate_block_hash_extra_keys, thus commented out
    # if mm_positions and len(mm_positions) != len(mm_hashes):
    #    raise ValueError(
    #        "The number of multi-modal positions and hashes must match.")
    
    # We can also make this check include LoRA
    exist_mm_data = bool(mm_positions)
    curr_mm_idx = 0
    extra_keys = None

    ret = []
    parent_block_hash_value = None
    for start in range(0, len(token_ids), block_size):
        end = start + block_size
        block_token_ids = tuple(token_ids[start:end])
        # Do not hash the block if it is not full.
        if len(block_token_ids) < block_size:
            break

        # Update extra_keys and curr_mm_idx if the block involves
        # multimodal data 
        if exist_mm_data: 
            extra_keys, curr_mm_idx = generate_block_hash_extra_keys(
                request, start, end, curr_mm_idx)

        block_hash = hash_block_tokens(parent_block_hash_value,
                                       block_token_ids)
                                       block_token_ids, extra_keys)
    
```
This way the code itself indicates the clear separation when extra keys are needed for token hash.